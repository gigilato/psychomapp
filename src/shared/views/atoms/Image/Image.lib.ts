import { has, get } from 'lodash'
import { useCallback, useEffect, useState } from 'react'
import { Image as RNImage } from 'react-native'

import { images } from '$assets/images'

import { ImageSource } from './Image.props'

export const getImageSource = (source: ImageSource) => {
  if (typeof source === 'string' && has(images, source)) return get(images, source)
  return source
}

export const useRatioImage = (source: ImageSource, width?: number, height?: number) => {
  if (!width && !height) throw new Error('You must provide at least one of width or height')

  const [ratioStyle, setRatioStyle] = useState({ width, height })
  const generateStyle = useCallback(
    (realWidth: number, realHeight: number) => {
      const isWidth = !!width
      const value = isWidth ? width! : height!
      const ratioValue = isWidth
        ? (value * realHeight) / realWidth
        : (value * realWidth) / realHeight
      setRatioStyle({ width: isWidth ? value : ratioValue, height: isWidth ? ratioValue : value })
    },
    [height, width]
  )

  useEffect(() => {
    const resolvedSource = getImageSource(source)
    if (typeof resolvedSource === 'number') {
      const { width: imageRealWidth, height: imageRealHeight } =
        RNImage.resolveAssetSource(resolvedSource)
      generateStyle(imageRealWidth, imageRealHeight)
    } else if (typeof resolvedSource === 'object') {
      if (resolvedSource.headers)
        RNImage.getSizeWithHeaders(
          resolvedSource.uri,
          resolvedSource.headers,
          (imageRealWidth, imageRealHeight) => generateStyle(imageRealWidth, imageRealHeight)
        )
      else
        RNImage.getSize(resolvedSource.uri, (imageRealWidth, imageRealHeight) =>
          generateStyle(imageRealWidth, imageRealHeight)
        )
    }
  }, [height, width, source, generateStyle])
  return ratioStyle
}
