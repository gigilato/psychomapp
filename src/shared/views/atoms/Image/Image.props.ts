import { ImageProps, ImageSource as DefaultImageSource } from 'expo-image'

import { images } from '$assets/images'
import { LiteralUnion } from '$types/common'

export type ImageSource = LiteralUnion<keyof typeof images> | number | DefaultImageSource

export interface IImageProps extends ImageProps {
  source: ImageSource
  width?: number
  height?: number
}
