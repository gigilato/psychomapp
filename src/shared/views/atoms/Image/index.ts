export * from './Image'
export * from './Image.props'
