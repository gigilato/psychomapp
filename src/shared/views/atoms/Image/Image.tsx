import { Image as ExpoImage } from 'expo-image'
import { styled } from 'nativewind'
import { memo } from 'react'

import { getImageSource, useRatioImage } from './Image.lib'
import { IImageProps } from './Image.props'

const StyledImage = styled(ExpoImage)
const DefaultImage = ({ source, ...props }: IImageProps) => {
  return <StyledImage source={getImageSource(source)} {...props} />
}

const RatioImage = memo<IImageProps>(({ source, width, height, style, ...props }) => {
  const ratioStyle = useRatioImage(source, width, height)
  return <Image source={source} style={[ratioStyle, style]} contentFit="contain" {...props} />
})

export const Image = ({ width, height, ...props }: IImageProps) => {
  if (width && height) return <RatioImage width={width} height={height} {...props} />

  return <DefaultImage {...props} />
}
