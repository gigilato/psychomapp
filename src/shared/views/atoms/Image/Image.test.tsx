import { render } from '@testing-library/react-native'

import { Image } from './Image'

describe('Image', () => {
  it('renders with contain', async () => {
    const component = render(<Image source="logo" className="h-[150]" contentFit="contain" />)
    expect(component).toMatchSnapshot()
  })

  it('renders with cover', async () => {
    const component = render(<Image source="logo" className="h-[150]" contentFit="cover" />)
    expect(component).toMatchSnapshot()
  })

  it('renders with ratio', async () => {
    const component = render(<Image source="logo" width={100} />)
    expect(component).toMatchSnapshot()
  })
})
