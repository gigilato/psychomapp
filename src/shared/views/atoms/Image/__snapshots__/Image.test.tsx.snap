// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders with contain 1`] = `
<ViewManagerAdapter_ExpoImage
  contentFit="contain"
  contentPosition={
    {
      "left": "50%",
      "top": "50%",
    }
  }
  height={150}
  onError={[Function]}
  onLoad={[Function]}
  onLoadStart={[Function]}
  onProgress={[Function]}
  placeholder={[]}
  source={[]}
  style={
    {
      "height": 150,
    }
  }
  transition={null}
/>
`;

exports[`Image renders with cover 1`] = `
<ViewManagerAdapter_ExpoImage
  contentFit="cover"
  contentPosition={
    {
      "left": "50%",
      "top": "50%",
    }
  }
  height={150}
  onError={[Function]}
  onLoad={[Function]}
  onLoadStart={[Function]}
  onProgress={[Function]}
  placeholder={[]}
  source={[]}
  style={
    {
      "height": 150,
    }
  }
  transition={null}
/>
`;

exports[`RatioImage renders correctly 1`] = `
<ViewManagerAdapter_ExpoImage
  contentFit="contain"
  contentPosition={
    {
      "left": "50%",
      "top": "50%",
    }
  }
  height={100}
  onError={[Function]}
  onLoad={[Function]}
  onLoadStart={[Function]}
  onProgress={[Function]}
  placeholder={[]}
  source={[]}
  style={
    {
      "height": 100,
      "width": 100,
    }
  }
  transition={null}
  width={100}
/>
`;
